swupd/delta.go

```diff
@@ -0,0 +1,190 @@
+package swupd
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+	"sync"
+
+	"github.com/clearlinux/mixer-tools/helpers"
+	"github.com/pkg/errors"
+)
+
+// CreateDeltas creates all delta files between the previous and current
+// version of the supplied manifest, returning a list of files that
+// delta creation failed for.
+func CreateDeltas(manifest, statedir string, from, to uint32) ([]*File, error) {
+	var c config
+
+	c, err := getConfig(statedir)
+	if err != nil {
+		return nil, err
+	}
+
+	var oldManifest *Manifest
+	var newManifest *Manifest
+
+	if oldManifest, err = ParseManifestFile(filepath.Join(c.outputDir, fmt.Sprintf("%d", from), manifest)); err != nil {
+		return nil, err
+	}
+	if newManifest, err = ParseManifestFile(filepath.Join(c.outputDir, fmt.Sprintf("%d", to), manifest)); err != nil {
+		return nil, err
+	}
+
+	// Must be sorted before passing into linkPeersAndChange()
+	oldManifest.sortFilesName()
+	newManifest.sortFilesName()
+
+	_, _, _ = newManifest.linkPeersAndChange(oldManifest, from)
+
+	deltas, err := consolidateDeltaFiles(newManifest, from, c)
+	if err != nil {
+		return nil, errors.Wrapf(err, "Failed to create deltas list %s", manifest)
+	}
+
+	const maxRoutines = 3
+	var failed []*File
+	var deltaQueue = make(chan *File)
+	var failedChan = make(chan *File, maxRoutines)
+	var wg sync.WaitGroup
+	wg.Add(maxRoutines)
+
+	for i := 0; i < maxRoutines; i++ {
+		go func(statedir string) {
+			defer wg.Done()
+
+			for f := range deltaQueue {
+				err := createDelta(f, statedir, f.DeltaPeer.Version, f.DeltaPeer.Hash)
+				if err != nil {
+					failedChan <- f
+				}
+			}
+		}(statedir)
+	}
+
+	for _, file := range deltas {
+		select {
+		case deltaQueue <- file:
+		case f := <-failedChan:
+			failed = append(failed, f)
+		}
+	}
+
+	close(deltaQueue)
+	wg.Wait()
+
+	if len(failedChan) > 0 {
+		f := <-failedChan
+		failed = append(failed, f)
+	}
+
+	return failed, nil
+}
+
+func consolidateDeltaFiles(manifest *Manifest, from uint32, c config) ([]*File, error) {
+	var files []*File
+
+	if manifest == nil {
+		return nil, nil
+	}
+
+	for _, file := range manifest.Files {
+		if file.Version <= from {
+			continue
+		}
+		if file.DeltaPeer == nil {
+			continue
+		}
+		if file.Type != TypeFile ||
+			file.DeltaPeer.Type != TypeFile {
+			continue
+		}
+
```

Comment:
Should the check for Status Deleted/Ghosted and the check whether is same file happen here? Otherwise we are generating a list of files to create deltas that has files that might not need delta at all.

Reply:
The check shouldn't actually be needed anymore. If you look at the latest `linkPeersAndChange` code delta peers will only be set on files with the same name that are not the same file and (I believe) are not ghosted or deleted.